core/src/main/java/com/sequenceiq/cloudbreak/controller/StackCreatorService.java

```diff
@@ -370,9 +372,13 @@ private void setStackTypeAndValidateDatalake(Stack stack, Blueprint blueprint) {
         }
     }
 
-    private void fillInstanceMetadata(Stack stack) {
+    void fillInstanceMetadata(Stack stack) {
         long privateIdNumber = 0;
-        for (InstanceGroup instanceGroup : stack.getInstanceGroups()) {
+        //Gateway HostGroups are sorted first to start with privateIdNumber 0.
+        List<InstanceGroup> sortedInstanceGroups = stack.getInstanceGroups().stream()
+                .sorted(Comparator.comparing(InstanceGroup::getInstanceGroupType)
```

Comment:
I don't think this will provide the Knox node(s) (which effectively is the endpoint for all services) with a consistent hostname. e.g. If the number of 'CORE' nodes is dynamic - the 'GATEWAY' instance will get a number which follows the count of all 'CORE' nodes.
This needs better understanding of 'CORE' vs 'GATEWAY'. If I'm not mistaken, there can be one GATEWAY hostGroup, and this is the instance on which KNOX is installed (not sure if there can be more than one such node). I could be completely wrong about this though. @cegganesh84 - I think you have a much better understanding of CORE vs GATEWAY and cardinality given the HA templates that you've worked on.
If GATEWAY==KNOX node and there can be only 1 gateway hostGroup, explicitly assigning numbers to this hostGroup first probably works.
Alternately, may need to make this service aware.

Reply:
Looking some more, looks like GATEWAY is the first in the sort order, so this works as expected. May still be better to explicitly look at GATEWAY nodes first.
Still need to get an understanding of where Knox goes, and how many GATEWAY type groups can exist.

Reply:
Knox is always deployed to the CM's machine which group always have the `GATEWAY` as InstanceGroupType. In case of HA clusters where could be multiple gateway instance/host group due to the comparison on group name the private id selection should be consistent - I think

Reply:
Previously, we had multiple groups with GATEWAY type for Ambari HA. At some point, the same will come to CM HA as well.

Reply:
Yes, if multiple GATEWAY host-groups are defined in future (current template validation fails multiple gateway host-groups), the naming convention will still be consistent for the same template due to comparison on InstanceGroupType which is followed by groupName. so nodes will masterA0, masterB1  if masterA, masterB are gateway hostgroups.

Also tested cluster launch by changing gateway hostGroup from "master"  to "worker"  and then "worker" nodes started with worker0.

CB-6185 concern is basically about different\random node sequenceId and hence node name for same template and cardinality and that issue would be addressed by this patch .